Set a default presentation pattern for String/Number/Datetime items

lolodomo opened the pull request:
A default state pattern was previously provided by ChannelStateDescriptionProvider only for String and Number items linked to a channel.
It is now the class DefaultStateDescriptionFragmentProvider which is responsible for providing the default state pattern for items, whether the item is linked to a channel or not.
This new class is the lowest ranked StateDescriptionFragmentProvider so that all other providers have priority in setting the state pattern.

Default pattern for string item: %s
Default pattern for datetime item or group with datetime state: %1$tY-%1$tm-%1$td %1$tH:%1$tM:%1$tS
Default pattern for number item or group with number state: %.0f
Default pattern for number+dimension item or group with number+dimension state: %.0f %unit%

Closes #4071
Closes #3835

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

lolodomo commented:
With the following items:
```
String DemoString
Number DemoNumber
DateTime DemoDateTime

String DemoString2 "DemoString2 [* %s]"
String DemoString3 "DemoString3 []"
Number DemoNumber2 "DemoNumber2 [%f]"
Number DemoNumber3 "DemoNumber3 []"
DateTime DemoDateTime2 "DemoDateTime2 [%1$td.%1$tm.%1$tY]"
DateTime DemoDateTime3 "DemoDateTime3 []"
```
and the following portion of sitemap:
```
	Text label="Default pattern" {
            Text item=DemoString
            Text item=DemoDateTime
            Text item=DemoNumber
            Text item=DemoString2
            Text item=DemoDateTime2
            Text item=DemoNumber2
            Text item=DemoString3
            Text item=DemoDateTime3
            Text item=DemoNumber3
	}
```

The three elements are items with no pattern defined and you can see that a default pattern is now applied:

![image](https://github.com/openhab/openhab-core/assets/10186704/d0365f95-eb84-4fad-a30d-b699080f179b)

The result without this change:

![image](https://github.com/openhab/openhab-core/assets/10186704/ee6729a7-de68-4a35-9b0e-58d3965822db)

J-N-K commented:
Why should there be another default than what we have now? It has always been argued that "textual configuration" offers finer control for advanced users. It is already possible to configure state descriptions, whoever wants that, can do it.

lolodomo commented:
> Why should there be another default than what we have now?

There is no default, that is the subject.

> It has always been argued that "textual configuration" offers finer control for advanced users. It is already possible to configure state descriptions, whoever wants that, can do it.

That is a good argument.
Having a default would just avoids defining it everytime. See that as a way to simplify config files.

mherwege commented:
I am just wondering if it shouldn't be implemented here: https://github.com/openhab/openhab-core/blob/cbb458e0c3c35f353954f82031c5cd1b3943759c/bundles/org.openhab.core/src/main/java/org/openhab/core/internal/service/StateDescriptionServiceImpl.java#L64
That would keep the behaviour between items in UI and config files consistent. That doesn't mean the UI shouldn't be enhanced to give some more guidance when configuring, but I don't like the inconsistency in behaviour.

lolodomo commented:
> Did you check what happens if there are state options provided? Would setting a pattern have an impact?

You can't provide state options through .items file.
The place I changed is the same where is built the state pattern when one is provided in the label in .items file.

mherwege commented:
> You can't provide state options through .items file.

I think you can (provided as metadata). See: https://community.openhab.org/t/oh3-add-metadata-to-items-via-configuration-files/112843

lolodomo commented:
Adding a default pattern like I did would have the same effect as setting explicitly this pattern in the items file.

mherwege commented:
> Adding a default pattern like I did would have the same effect as setting explicitly this pattern in the items file.

Correct. And that is my point. Before it was possible to have no pattern at all. My fear is that the default pattern you will now get interferes with configuration done somewhere else in a metadata state description, as both will get merged. And the merge is not very smart if I understand the code well, it will start from the first it finds and gradually merge other state description sources. I don't know if it is an issue, but I think it should be checked.

lolodomo commented:
If my reading of the code is correct, if a pattern is already provided by MetadataStateDescriptionFragmentProvider, the default or explicitly set pattern provided by GenericItemProvider will be ignored.

I will confirm the order of providers in getMergedStateDescriptionFragments.

lolodomo commented:
What could be more random even in the current code is between ChannelStateDescriptionProvider and GenericItemProvider having I believe same rank 0. If confirmed, the current code could not guarantee what takes precedence between both of them.

ChannelStateDescriptionProvider should have lower priority.
Maybe we should have rank 0 for ChannelStateDescriptionProvider, rank 1 for GenericItemProvider and rank 2 for MetadataStateDescriptionFragmentProvider ?

lolodomo commented:
@mherwege : I believe it is clear for me now what should be a proper and generic solution. I am going to create a new StateDescriptionFragmentProvider in charge to provide the default pattern for each item. I will set rank for it at lowest value so that this default pattern will be used only if not provided by any other StateDescriptionFragmentProvider.

lolodomo commented:
> What could be more random even in the current code is between ChannelStateDescriptionProvider and GenericItemProvider having I believe same rank 0. If confirmed, the current code could not guarantee what takes precedence between both of them.

At runtime, ChannelStateDescriptionProvider has rank -1 while GenericItemProvider has rank 0, so it is OK ... even if I don't understand where is defined this rank for ChannelStateDescriptionProvider.

lolodomo commented:
I updated my initial message for a better description of the change.
I updated my second message showing an example before this change and after this change.

PR is now ready for review.

lolodomo commented:
A little change in OH documentation will then be necessary to explain that there is now a default state pattern for certain item types if none is provided:
https://www.openhab.org/docs/configuration/items.html#state-presentation

lolodomo commented:
@openhab/core-maintainers : any chance to get a feedback ?

J-N-K commented on bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/DefaultStateDescriptionFragmentProvider.java:
```diff
@@ -0,0 +1,119 @@
+
+    private final Map<String, StateDescriptionFragment> stateDescriptionFragments = new ConcurrentHashMap<>();
+
+    private Integer rank = -2; // takes less precedence over all other providers
```
```suggestion
    private Integer rank = -2; // takes less precedence than all other providers
```

J-N-K commented on bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/DefaultStateDescriptionFragmentProvider.java:
```diff
@@ -0,0 +1,119 @@
+    @Activate
+    public DefaultStateDescriptionFragmentProvider(Map<String, Object> properties) {
+        Object serviceRanking = properties.get(Constants.SERVICE_RANKING);
+        if (serviceRanking instanceof Integer) {
+            rank = (Integer) serviceRanking;
+        }
```
```suggestion
        if (serviceRanking instanceof Integer intValue) {
            rank = intValue;
        }
```

J-N-K commented on bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/ItemRegistryImpl.java:
```diff
@@ -469,6 +485,22 @@ public void unsetCommandDescriptionService(CommandDescriptionService commandDesc
         }
     }
 
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
```
Wouldn't it make sense to make this a mandatory service and use constructor injection? It is a default and IMO should always be provided.

lolodomo replied:
You're right.

lolodomo commented:
@J-N-K : changes done. DCO is KO because I made a typo in my name in the commit just before the rebase to solve the conflict.

J-N-K commented on bundles/org.openhab.core/src/main/java/org/openhab/core/internal/items/DefaultStateDescriptionFragmentProvider.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    public void onItemAdded(Item item) {
+        logger.debug("DefaultStateDescriptionFragmentProvider onItemAdded {} {}", item.getName(), item.getType());
```
I would prefer to remove this log-statement, but if you want to keep them: `DefaultStateDescriptionFragmentProvider` is not needed, it's already part of the logger name and please reduce to TRACE to prevent log spamming. (also below)

lolodomo replied:
Done

J-N-K reviewed (changes requested):
Otherwise LGTM

lolodomo commented:
@J-N-K : can this one be merged or is there still something to adjust ?
